[Snippets][Docs] MHA Optimization Guide (#25493)
openvinotoolkit/openvino: src/common/snippets/docs/mha_optimization_guide.md

Author: Thanks a lot for pointing it out! I haven't worked much with .md files before, so this is a completely new thing for me. I'll refactor the file in accordance to the suggestion

Author: Good idea, thank you! I added information about supported Transpose layers in `ExtractUnsupportedTransposes` section: the `FuseTransposeBrgemm` pass is mentioned there. Please check

The CPU plugin callback for TokenizeMHASnippets is placed in [transformation_pipeline.cpp](../../../plugins/intel_cpu/src/transformations/transformation_pipeline.cpp) file (please see the code in `MainSnippets` method).

**Please note that the CPU callback is usually ignored in cpu functional tests: SnippetsMode::IgnoreCallback is used for that**.
Currently, SnippetsMode has 3 states: `Enable`, `IgnoreCallback` and `Disable`.
For the details, please refer to [ov::intel_cpu::Config](../../../plugins/intel_cpu/src/config.h).

## Snippets Common Optimizations

After tokenization, snippets [common optimizations](../src/pass/common_optimizations.cpp) are applied to the tokenized Subgraphs.
These transformations can modify both the Subgraph's body and its surroundings (e.g. extract constant nodes outside the Subgraph).
Let's explore several transformations that can impact MHA performance.

### ExtractUnsupportedTransposes

[ExtractUnsupportedTransposes](../src/pass/extract_unsupported_transposes.cpp) moves up unsupported Transposes outside the Subgraph.

Snippets support 2 types of Transposes:

> 1. Transposes on Matmul inputs and output: these Transposes will be fused into Brgemm (which supports strided read/write) node by [FuseTransposeBrgemm](../src/pass/fuse_transpose_brgemm.cpp) data flow transformation.

Reviewer: According to the diagram you provided, all Transposes are on Matmult's inputs or outputs, so this distinction is a bit confusing.
I think the main criteria, that that the transpose doesn't change the last dimension => can be executed inplace by specifying strides. It's already mentioned, let's just remove the first part before `:` and repharase a little

Author: Done, thanks!

The supported Transpose orders for Brgemm fusion are defined by `TokenizeMHASnippets::get_fusion_transpose_order` in [mha_tokenization.cpp](../src/pass/mha_tokenization.cpp)

> 2. The rest transposes are decomposed by [TransposeDecomposition](../src/pass/transpose_decomposition.cpp) data flow transformation.

Reviewer: ```suggestion
2. The rest of transposes are decomposed by [TransposeDecomposition](../src/pass/transpose_decomposition.cpp) data flow transformation.
```

## MHA Tokenization

This structure represents the basic MHA pattern that can be tokenized by Snippets MHA pattern:

```mermaid
graph TB
    MM0A[Transpose/Eltwise] --> MatMul0
    MM0B[Transpose/Eltwise] --> MatMul0
    MatMul0 --> IntermediateBeforeSM[Transpose/Eltwise/Select/Reshape/FakeQuantize]
    IntermediateBeforeSM --> Softmax
    Softmax --> IntermediateAfterSM[Transpose/Eltwise/Select/Reshape/FakeQuantize]
    IntermediateAfterSM --> MatMul1
    MM1B[Transpose/Eltwise] --> MatMul1
    MatMul1 --> Transpose/Eltwise
```

Please note, that layers, denoted by `/`, can represent both single nodes and sequences of nodes. The code, which performs the tokenization, is placed in [TokenizeMHASnippets](../src/pass/mha_tokenization.cpp) transformation.

### CPU Plugin Callback for MHA Tokenization

The tokenization can be regulated via callback. In CPU plugin, the callback prevents tokenization in certain cases, which can be categorized into two types:

> 1. Pattern unsupported by Snippets CPU backend.

Author: Thanks for the suggestion, applied

2. Tokenization restrictions for performance reasons: executing MHA as separated operations may be faster in some cases.

The CPU plugin callback for TokenizeMHASnippets is placed in [transformation_pipeline.cpp](../../../plugins/intel_cpu/src/transformations/transformation_pipeline.cpp) file (please see the code in `MainSnippets` method).

> **Please note that the CPU callback is usually ignored in cpu functional tests.**

Author: Updated, please check

## Snippets Common Optimizations

After tokenization, snippets [common optimizations](../src/pass/common_optimizations.cpp) are applied to the tokenized Subgraphs. These transformations can modify both the Subgraph's body and its surroundings (e.g. extract constant nodes outside the Subgraph). Let's explore several transformations that can impact MHA performance.

### ExtractUnsupportedTransposes

> [ExtractUnsupportedTransposes](../src/pass/extract_unsupported_transposes.cpp) moves up unsupported Transposes outside the Subgraph. **Please note: the "unsupported" Transpose actually can be executed via Snippets, however CPU plugin implementation usually works faster.**

Author: I added an explanation on which transposes are supported, so the rest ones are affected by the transformation. I think it is clear enough, but feel free to share your concerns

### SplitDimensionM

> [SplitDimensionM](../src/pass/split_dimension_m.cpp) splits M dimension of MHA on 2 parts (`batch_m` and `new_m`) by inserting Reshape on A input of the first Matmul and output of the second Matmul (the rest Subgraph's inputs are reshaped by Unsqueeze-like reshapes in order not to break subgraph semantic). This allows to increase parallel work amount due to `batch_m`. The splitting is performed based on heuristic algorithm which can be found in `SplitDimensionM::get_splited_dimensions` method.

Author: Thanks for the idea about picture, I added it. Since we have the picture now, I think it's obvious from it now what "Unsqueeze-like reshapes" are.

**Important notes:**
- Since `SplitDimensionM` depends on parallel concurrency (number of threads which are used in one stream), the transformation result depends not only on the HW platform, but on number of streams used during model inference as well. For instance, this might lead to different result in throughput and latency hint modes.
- `SplitDimensionM::can_be_optimized` is used in CPU plugin callback: if this method reports that appropriate parallel work amount can not be set for the MHA, the tokenization doesn't happen.

## Brgemm Blocking

Within the Snippets CPU backend, the MatMul is executed using the Brgemm primitive. For enhancing the execution efficiency, blocking across the M, K, and N matmul dimensions is used.

### Blocking Parameters

The heuristics for determining the optimal block sizes can be found in [SetBrgemmCPUBlockingParams](../../../plugins/intel_cpu/src/transformations/snippets/x64/pass/set_brgemm_cpu_blocking_params.cpp).

**Important notes:**
- Blocking by M dimension is shared between both Brgemms. Please see [SplitLoops](../include/snippets/lowered/pass/split_loops.hpp) lowered pass for the details.

> - The heuristics from `SetBrgemmCPUBlockingParams` are not currently applied by K and N dimensions for BF16/INT8 Brgemms, although it is technically supported.

Reviewer: I don't think that we need to point out all todos and technical limitations of current pipeline. This would not make the docs any more readable.
Let's remove this bullet (The heuristics...) and merge the previous one with the paragraph before Important Notes

Author: Done, please check

### Blocking Order

The lowered pass [BrgemmBlocking](../../../plugins/intel_cpu/src/transformations/snippets/x64/pass/lowered/brgemm_blocking.cpp) performs blocking loops creation on LinearIR. Currently, the order of blocking loops is following (from outer to inner): `M->N->K`.

## MHA Performance Tuning Recommendations

> Buiding on the information presented upper, this section focuses on providing recommendations for fine-tuning the performance of specific MHA pattern.

Reviewer: ```suggestion
Based on previously discussed information, we provide the following recommendations for the MHA performance fine-tuning:
```

1. Check if there are MHA's which were not tokenized because of [CPU plugin callback](#cpu-plugin-callback-for-mha-tokenization).
2. Check how the graph was changed by [CommonOptimizations](#snippets-common-optimizations). In local experiments, some transformations might be worth to change:
    - Disable [ExtractUnsupportedTransposes](#extractunsupportedtransposes) transformation in order to benchmark Snippets Transpose implementation.
    - Adjust [SplitDimensionM](#splitdimensionm) heuristics in order to benchmark another splitting, or disable the pass at all.
3. [Blocking parameters](#blocking-parameters): adjust blocking heuristics in `SetBrgemmCPUBlockingParams`.
    - Please note that there is 2 Matmul nodes inside a single MHA, and each Matmul can have his own optimal K, N blocking params. M block is better to keep the same since the corresponding blocking loop is shared between both Matmuls.
    - For the BF16/INT8 blocking loops, 2 options are possible: blocking can be done only for Brgemm node, or for BrgemmCopyB repacking too.

> 4. [Blocking order](#blocking-order): blocking order changing might have an effect as well. But please note, that blocking by M dimensions is shared between both Brgemms only if blocking by M is outermost.

Reviewer: Not sure that we support changing of blocking order out-of-the box. I would rather skip this point.

Author: Okay, let's remove this point

> By following these recommendations, signifficant performance improvement for the specific MHA can be achieved. Additionally, the results of these experiments can be used as a solid foundation for the subsequent heuristics adjusting.

Reviewer: > signifficant performance improvement for the specific MHA can be achieved.
After reading this sentence, one might think that we don't provide adequate out-of-the box performance for most of the patterns 🙂 
Let's paraphrase a little bit:
```suggestion
Following these recommendations, the performance of some specific MHA patters can be fine-tuned. Additionally, the results of these experiments can be used as a solid foundation for the subsequent heuristics adjustments. 
```

# MHA Optimization Guide

## Introduction

This guide explores the mechanism of the Multi Head Attention (MHA) patterns tokenization and several methods that are used for MHA performance optimization. Also, there is provided several recommendations on how to fine-tune performance of the specific MHA pattern.

## MHA Tokenization

This structure represents the basic MHA pattern that can be tokenized by Snippets MHA pattern:

```mermaid
graph TB

>     MM0A[Transpose/Eltwise] --> MatMul0

Author: Thank you! The scheme is actualized

    MM0B[Transpose/Eltwise] --> MatMul0
    MatMul0 --> IntermediateBeforeSM[Transpose/Eltwise/Select/Reshape/FakeQuantize]
    IntermediateBeforeSM --> Softmax
    Softmax --> IntermediateAfterSM[Transpose/Eltwise/Select/Reshape/FakeQuantize]
    IntermediateAfterSM --> MatMul1
    MM1B[Transpose/Eltwise] --> MatMul1

>     MatMul1 --> Transpose/Eltwise

Reviewer 2: Also `FakeQuantize` can be after `MatMul1` also

>     MatMul1 --> Transpose/Eltwise

Author: It's quite difficult to cover all the details in the scheme and don't overload it :)
I think the current version is enough for general representation, and all the details can be found in the code. If you disagree, please let me know: I will think how to appropriately update the picture.
As for `FakeQuantize` on MM1 output: I updated the scheme, thanks

>     MatMul1 --> Transpose/Eltwise

Reviewer 2: >It's quite difficult to cover all the details in the scheme and don't overload it :)
I fully agreed with you! When I saw `Please note, that layers, denoted by "/", can represent both single nodes and sequences of nodes.`, as the author of this non-straightforward transformation, I was confused a little 😄 
However, let's leave it so -> if needed, the developer can take a look at the pass. You're right!
